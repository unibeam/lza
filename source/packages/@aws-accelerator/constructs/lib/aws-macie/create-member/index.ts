/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import { getGlobalRegion } from '@aws-accelerator/utils/lib/common-functions';
import * as AWS from 'aws-sdk';
AWS.config.logger = console;

/**
 * add-macie-members - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
      StatusCode: number | undefined;
    }
  | undefined
> {
  const region = event.ResourceProperties['region'];
  const partition = event.ResourceProperties['partition'];
  const adminAccountId = event.ResourceProperties['adminAccountId'];
  const solutionId = process.env['SOLUTION_ID'];
  const globalRegion = getGlobalRegion(partition);
  const organizationsClient = new AWS.Organizations({ customUserAgent: solutionId, region: globalRegion });
  const macie2Client = new AWS.Macie2({ region: region, customUserAgent: solutionId });
  const allAccounts: AWS.Organizations.Account[] = [];
  const existingMembers: AWS.Macie2.Member[] = [];
  let isEnabled = false;

  let nextToken: string | undefined = undefined;
  do {
    const page = await throttlingBackOff(() => organizationsClient.listAccounts({ NextToken: nextToken }).promise());
    for (const account of page.Accounts ?? []) {
      allAccounts.push(account);
    }
    nextToken = page.NextToken;
  } while (nextToken);

  nextToken = undefined;
  do {
    const page = await throttlingBackOff(() =>
      macie2Client.listMembers({ nextToken, onlyAssociated: 'false' }).promise(),
    );
    for (const member of page.members ?? []) {
      existingMembers.push(member);
    }
    nextToken = page.nextToken;
  } while (nextToken);

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      if (!(await isMacieEnable(macie2Client))) {
        console.log('start enable of macie');
        await throttlingBackOff(() =>
          macie2Client
            .enableMacie({
              status: 'ENABLED',
            })
            .promise(),
        );
      }

      for (const account of allAccounts.filter(item => item.Id !== adminAccountId) ?? []) {
        const existingMember = existingMembers.find(member => member.accountId !== account.Id);
        if (existingMember && existingMember.relationshipStatus === 'Removed') {
          console.log(
            `OU account - ${account.Id} macie membership status is "Removed", deleting member before adding again`,
          );
          await throttlingBackOff(() => macie2Client.deleteMember({ id: account.Id! }).promise());
        }
        if (!existingMember || existingMember.relationshipStatus === 'Removed') {
          console.log(`OU account - ${account.Id} macie membership status is "not a macie member", adding as a member`);
          await throttlingBackOff(() =>
            macie2Client
              .createMember({
                account: { accountId: account.Id!, email: account.Email! },
              })
              .promise(),
          );
        } else {
          console.warn(
            `OU account - ${account.Id} macie membership status is "a macie member", ignoring create member for the account!!`,
          );
        }
      }

      isEnabled = await isOrganizationAutoEnabled(macie2Client);

      if (!isEnabled) {
        await throttlingBackOff(() => macie2Client.updateOrganizationConfiguration({ autoEnable: true }).promise());
      } else {
        console.warn('Delegation admin account Auto-Enable is ON, so ignoring');
      }
      return { Status: 'Success', StatusCode: 200 };
    case 'Delete':
      for (const account of allAccounts.filter(item => item.Id !== adminAccountId) ?? []) {
        if (existingMembers.find(member => member.accountId !== account.Id)) {
          console.log(
            `OU account - ${account.Id} macie membership status is "a macie member", removing from member list`,
          );
          await throttlingBackOff(() => macie2Client.disassociateMember({ id: account.Id! }).promise());
          await throttlingBackOff(() => macie2Client.deleteMember({ id: account.Id! }).promise());
        } else {
          console.warn(
            `OU account - ${account.Id} macie membership status is "not a macie member", ignoring removing member list!!`,
          );
        }
      }

      isEnabled = await isOrganizationAutoEnabled(macie2Client);
      if (isEnabled) {
        await throttlingBackOff(() => macie2Client.updateOrganizationConfiguration({ autoEnable: false }).promise());
      } else {
        console.warn('Delegation admin account Auto-Enable is OFF, so ignoring');
      }
      return { Status: 'Success', StatusCode: 200 };
  }
}

/**
 * Checking is organization auto enabled for new account
 * @param macieClient
 */
async function isOrganizationAutoEnabled(macieClient: AWS.Macie2): Promise<boolean> {
  console.log('calling isOrganizationAutoEnabled');
  const response = await throttlingBackOff(() => macieClient.describeOrganizationConfiguration({}).promise());
  return response.autoEnable ?? false;
}

/**
 * Checking Macie is enable or disabled
 * @param macie2Client
 */
async function isMacieEnable(macie2Client: AWS.Macie2): Promise<boolean> {
  try {
    const response = await throttlingBackOff(() => macie2Client.getMacieSession({}).promise());
    return response.status === 'ENABLED';
  } catch (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    e: any
  ) {
    if (
      // SDKv2 Error Structure
      e.code === 'ResourceConflictException' ||
      // SDKv3 Error Structure
      e.name === 'ResourceConflictException'
    ) {
      console.warn(e.name + ': ' + e.message);
      return false;
    }

    // This is required when macie is not enabled AccessDeniedException exception issues
    if (
      // SDKv2 Error Structure
      e.code === 'AccessDeniedException' ||
      // SDKv3 Error Structure
      e.name === 'AccessDeniedException'
    ) {
      console.warn(e.name + ': ' + e.message);
      return false;
    }
    throw new Error(`Macie enable issue error message - ${e}`);
  }
}
