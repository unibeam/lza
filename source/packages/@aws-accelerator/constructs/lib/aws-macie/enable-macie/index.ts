/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import * as AWS from 'aws-sdk';
AWS.config.logger = console;

/**
 * add-macie-members - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
      StatusCode: number | undefined;
    }
  | undefined
> {
  const region = event.ResourceProperties['region'];
  const findingPublishingFrequency = event.ResourceProperties['findingPublishingFrequency'];
  const isSensitiveSh = event.ResourceProperties['isSensitiveSh'] === 'true';
  const solutionId = process.env['SOLUTION_ID'];

  const macie2Client = new AWS.Macie2({ region: region, customUserAgent: solutionId });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      let macieStatus = await isMacieEnable(macie2Client);
      if (!macieStatus) {
        console.log('start enable of macie');
        await throttlingBackOff(() =>
          macie2Client
            .enableMacie({
              findingPublishingFrequency: findingPublishingFrequency,
              status: 'ENABLED',
            })
            .promise(),
        );
      }
      console.log('start update of macie');
      await throttlingBackOff(() =>
        macie2Client
          .updateMacieSession({
            findingPublishingFrequency: findingPublishingFrequency,
            status: 'ENABLED',
          })
          .promise(),
      );

      // macie status do not change immediately causing failure to other processes, so wait till macie enabled
      while (!macieStatus) {
        console.log(`checking macie status ${macieStatus}`);
        macieStatus = await isMacieEnable(macie2Client);
      }

      await throttlingBackOff(() =>
        macie2Client
          .putFindingsPublicationConfiguration({
            securityHubConfiguration: {
              publishClassificationFindings: isSensitiveSh,
              publishPolicyFindings: true,
            },
          })
          .promise(),
      );

      return { Status: 'Success', StatusCode: 200 };

    case 'Delete':
      if (await isMacieEnable(macie2Client)) {
        await throttlingBackOff(() =>
          macie2Client
            .disableMacie({
              findingPublishingFrequency: findingPublishingFrequency,
              status: 'ENABLED',
            })
            .promise(),
        );
      }
      return { Status: 'Success', StatusCode: 200 };
  }
}

async function isMacieEnable(macie2Client: AWS.Macie2): Promise<boolean> {
  try {
    const response = await throttlingBackOff(() => macie2Client.getMacieSession({}).promise());
    return response.status === 'ENABLED';
  } catch (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    e: any
  ) {
    if (
      // SDKv2 Error Structure
      e.code === 'ResourceConflictException' ||
      // SDKv3 Error Structure
      e.name === 'ResourceConflictException'
    ) {
      console.warn(e.name + ': ' + e.message);
      return false;
    }

    // This is required when macie is not enabled AccessDeniedException exception issues
    if (
      // SDKv2 Error Structure
      e.code === 'AccessDeniedException' ||
      // SDKv3 Error Structure
      e.name === 'AccessDeniedException'
    ) {
      console.warn(e.name + ': ' + e.message);
      return false;
    }
    throw new Error(`Macie enable issue error message - ${e}`);
  }
}
