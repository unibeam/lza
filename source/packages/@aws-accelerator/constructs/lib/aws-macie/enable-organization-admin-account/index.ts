/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { delay, throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import * as AWS from 'aws-sdk';
AWS.config.logger = console;

/**
 * enableOrganizationAdminAccount - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
      StatusCode: number | undefined;
    }
  | undefined
> {
  const adminAccountId = event.ResourceProperties['adminAccountId'];
  const region = event.ResourceProperties['region'];
  const solutionId = process.env['SOLUTION_ID'];

  const macie2Client = new AWS.Macie2({ region: region, customUserAgent: solutionId });

  const macieDelegatedAccount = await getMacieDelegatedAccount(macie2Client, adminAccountId);

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      // Enable macie in management account required to create delegated admin account
      let macieStatus = await isMacieEnable(macie2Client);
      if (!macieStatus) {
        console.log('start enable of macie');
        await throttlingBackOff(() =>
          macie2Client
            .enableMacie({
              status: 'ENABLED',
            })
            .promise(),
        );
      }

      // macie status do not change immediately causing failure to other processes, so wait till macie enabled
      while (!macieStatus) {
        console.log(`checking macie status ${macieStatus}`);
        macieStatus = await isMacieEnable(macie2Client);
      }

      if (macieDelegatedAccount.status) {
        if (macieDelegatedAccount.accountId === adminAccountId) {
          console.warn(
            `Macie admin account ${macieDelegatedAccount.accountId} is already an admin account as status is ${macieDelegatedAccount.status}, in ${region} region. No action needed`,
          );
          return { Status: 'Success', StatusCode: 200 };
        } else {
          console.warn(
            `Macie delegated admin is already set to ${macieDelegatedAccount.accountId} account can not assign another delegated account`,
          );
        }
      } else {
        console.log(
          `Started enableOrganizationAdminAccount function in ${event.ResourceProperties['region']} region for account ${adminAccountId}`,
        );
        let retries = 0;
        while (retries < 10) {
          await delay(retries ** 2 * 1000);
          try {
            await throttlingBackOff(() =>
              macie2Client.enableOrganizationAdminAccount({ adminAccountId: adminAccountId }).promise(),
            );
            break;
          } catch (error) {
            console.log(error);
            retries = retries + 1;
          }
        }
      }

      return { Status: 'Success', StatusCode: 200 };

    case 'Delete':
      if (macieDelegatedAccount.status) {
        console.log(
          `Started disableOrganizationAdminAccount function in ${event.ResourceProperties['region']} region for account ${adminAccountId}`,
        );
        await throttlingBackOff(() =>
          macie2Client.disableOrganizationAdminAccount({ adminAccountId: macieDelegatedAccount.accountId! }).promise(),
        );
      }

      return { Status: 'Success', StatusCode: 200 };
  }
}

async function getMacieDelegatedAccount(
  macie2Client: AWS.Macie2,
  adminAccountId: string,
): Promise<{ accountId: string | undefined; status: boolean }> {
  const adminAccounts: AWS.Macie2.AdminAccount[] = [];

  let nextToken: string | undefined = undefined;
  do {
    const page = await throttlingBackOff(() => macie2Client.listOrganizationAdminAccounts({ nextToken }).promise());
    for (const account of page.adminAccounts ?? []) {
      adminAccounts.push(account);
    }
    nextToken = page.nextToken;
  } while (nextToken);

  if (adminAccounts.length === 0) {
    return { accountId: undefined, status: false };
  }
  if (adminAccounts.length > 1) {
    throw new Error('Multiple admin accounts for Macie in organization');
  }

  if (adminAccounts[0].accountId === adminAccountId && adminAccounts[0].status === 'DISABLING_IN_PROGRESS') {
    throw new Error(`Admin account ${adminAccounts[0].accountId} is in ${adminAccounts[0].status}`);
  }

  return { accountId: adminAccounts[0].accountId, status: true };
}

async function isMacieEnable(macie2Client: AWS.Macie2): Promise<boolean> {
  try {
    const response = await throttlingBackOff(() => macie2Client.getMacieSession({}).promise());
    return response.status === 'ENABLED';
  } catch (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    e: any
  ) {
    if (
      // SDKv2 Error Structure
      e.code === 'ResourceConflictException' ||
      // SDKv3 Error Structure
      e.name === 'ResourceConflictException'
    ) {
      console.warn(e.name + ': ' + e.message);
      return false;
    }

    // This is required when macie is not enabled AccessDeniedException exception issues
    if (
      // SDKv2 Error Structure
      e.code === 'AccessDeniedException' ||
      // SDKv3 Error Structure
      e.name === 'AccessDeniedException'
    ) {
      console.warn(e.name + ': ' + e.message);
      return false;
    }
    throw new Error(`Macie enable issue error message - ${e}`);
  }
}
