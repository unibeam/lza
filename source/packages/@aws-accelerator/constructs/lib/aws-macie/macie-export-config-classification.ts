/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Initialized MacieExportConfigClassificationProps properties
 */
export interface MacieExportConfigClassificationProps {
  /**
   * Macie ExportConfigClassification repository bucket name
   */
  readonly bucketName: string;
  /**
   * Macie ExportConfigClassification repository bucket encryption key
   */
  readonly bucketKmsKey: cdk.aws_kms.IKey;
  /**
   * Bucket key prefix
   */
  readonly keyPrefix: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly logKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
  /**
   * Macie value for how frequently you want to publish the findings
   */
  readonly findingPublishingFrequency: 'FIFTEEN_MINUTES' | 'ONE_HOUR' | 'SIX_HOURS';

  /**
   * Macie value to determine if we publish classifications to Security Hub
   */
  readonly publishClassificationFindings: boolean;

  /**
   * Macie value to determine if we publish findings at all
   */
  readonly publishPolicyFindings: boolean;
}

/**
 * Aws MacieSession export configuration classification
 */
export class MacieExportConfigClassification extends Construct {
  public readonly id: string = '';

  constructor(scope: Construct, id: string, props: MacieExportConfigClassificationProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::MaciePutClassificationExportConfiguration';

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'put-export-config-classification/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Sid: 'MaciePutClassificationExportConfigurationTaskMacieActions',
          Effect: 'Allow',
          Action: [
            'macie2:EnableMacie',
            'macie2:GetClassificationExportConfiguration',
            'macie2:UpdateMacieSession',
            'macie2:GetMacieSession',
            'macie2:PutClassificationExportConfiguration',
            'macie2:PutFindingsPublicationConfiguration',
          ],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        region: cdk.Stack.of(this).region,
        bucketName: props.bucketName,
        keyPrefix: props.keyPrefix,
        kmsKeyArn: props.bucketKmsKey.keyArn,
        findingPublishingFrequency: props.findingPublishingFrequency,
        publishClassificationFindings: props.publishClassificationFindings,
        publishPolicyFindings: props.publishPolicyFindings,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.logKmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
