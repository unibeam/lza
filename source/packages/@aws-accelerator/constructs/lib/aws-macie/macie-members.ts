/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Initialized AwsMacieMembersProps properties
 */
export interface AwsMacieMembersProps {
  /**
   * Macie admin account id
   */
  readonly adminAccountId: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

/**
 /**
 * Class to Aws MacieSession Members
 */
export class MacieMembers extends Construct {
  public readonly id: string;

  constructor(scope: Construct, id: string, props: AwsMacieMembersProps) {
    super(scope, id);

    const MACIE_RESOURCE_TYPE = 'Custom::MacieCreateMember';

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, MACIE_RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'create-member/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Sid: 'MacieCreateMemberTaskOrganizationAction',
          Effect: 'Allow',
          Action: ['organizations:ListAccounts'],
          Resource: '*',
          Condition: {
            StringLikeIfExists: {
              'organizations:ListAccounts': ['macie.amazonaws.com'],
            },
          },
        },
        {
          Sid: 'MacieCreateMemberTaskMacieActions',
          Effect: 'Allow',
          Action: [
            'macie2:CreateMember',
            'macie2:DeleteMember',
            'macie2:DescribeOrganizationConfiguration',
            'macie2:DisassociateMember',
            'macie2:EnableMacie',
            'macie2:GetMacieSession',
            'macie2:ListMembers',
            'macie2:UpdateOrganizationConfiguration',
          ],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: MACIE_RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        region: cdk.Stack.of(this).region,
        partition: cdk.Stack.of(this).partition,
        adminAccountId: props.adminAccountId,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
