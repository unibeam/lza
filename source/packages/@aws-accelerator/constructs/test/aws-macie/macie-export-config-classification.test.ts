/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { MacieExportConfigClassification } from '../../lib/aws-macie/macie-export-config-classification';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(MacieExportConfigClassification): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new MacieExportConfigClassification(stack, 'AwsMacieUpdateExportConfigClassification', {
  publishPolicyFindings: true,
  findingPublishingFrequency: 'ONE_HOUR',
  publishClassificationFindings: true,
  bucketName: 'bucketName',
  keyPrefix: `111111111111-aws-macie-export-config`,
  logRetentionInDays: 3653,
  bucketKmsKey: new cdk.aws_kms.Key(stack, 'BucketKey', {}),
  logKmsKey: new cdk.aws_kms.Key(stack, 'LogKey', {}),
});

/**
 * MacieExportConfigClassification construct test
 */
describe('MacieExportConfigClassification', () => {
  snapShotTest(testNamePrefix, stack);
});
