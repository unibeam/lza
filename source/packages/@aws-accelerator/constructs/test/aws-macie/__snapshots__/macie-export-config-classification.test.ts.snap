// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MacieExportConfigClassification Construct(MacieExportConfigClassification):  Snapshot Test 1`] = `
{
  "Resources": {
    "AwsMacieUpdateExportConfigClassification832781E3": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomMaciePutClassificationExportConfigurationCustomResourceProviderLogGroup727354F4",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomMaciePutClassificationExportConfigurationCustomResourceProviderHandlerC53E2FCC",
            "Arn",
          ],
        },
        "bucketName": "bucketName",
        "findingPublishingFrequency": "ONE_HOUR",
        "keyPrefix": "111111111111-aws-macie-export-config",
        "kmsKeyArn": {
          "Fn::GetAtt": [
            "BucketKey7092080A",
            "Arn",
          ],
        },
        "publishClassificationFindings": true,
        "publishPolicyFindings": true,
        "region": {
          "Ref": "AWS::Region",
        },
      },
      "Type": "Custom::MaciePutClassificationExportConfiguration",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketKey7092080A": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomMaciePutClassificationExportConfigurationCustomResourceProviderHandlerC53E2FCC": {
      "DependsOn": [
        "CustomMaciePutClassificationExportConfigurationCustomResourceProviderRoleEB42D531",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomMaciePutClassificationExportConfigurationCustomResourceProviderRoleEB42D531",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomMaciePutClassificationExportConfigurationCustomResourceProviderLogGroup727354F4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "LogKeyE83D0C1B",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "CustomMaciePutClassificationExportConfigurationCustomResourceProviderHandlerC53E2FCC",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomMaciePutClassificationExportConfigurationCustomResourceProviderRoleEB42D531": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "macie2:EnableMacie",
                    "macie2:GetClassificationExportConfiguration",
                    "macie2:UpdateMacieSession",
                    "macie2:GetMacieSession",
                    "macie2:PutClassificationExportConfiguration",
                    "macie2:PutFindingsPublicationConfiguration",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "MaciePutClassificationExportConfigurationTaskMacieActions",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogKeyE83D0C1B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
